File: code/jvm/src/main/kotlin/pt/isel/daw/battleships/http/controllers/games/models/game/GameStateModel.kt
package pt.isel.daw.battleships.http.controllers.games.models.game

import pt.isel.daw.battleships.dtos.games.game.GameStateDTO
import javax.validation.constraints.Min
import javax.validation.constraints.Pattern

/**
 * Represents a game state.
 *
 * @property phase the current game phase
 * @property phaseEndTime the time when the current phase ends
 * @property round the current round
 * @property turn the current turn
 * @property winner the winner of the game
 */
data class GameStateModel(
    @Pattern(regexp = PHASE_REGEX) // TODO: This is Output, so no validation needed?
    val phase: String,
    val phaseEndTime: Long,

    @Min(value = MIN_ROUND.toLong(), message = "The round must be greater than or equal to $MIN_ROUND.")
    val round: Int?,
    val turn: String?,
    val winner: String?
) {
    constructor(stateDTO: GameStateDTO) : this(
        phase = stateDTO.phase,
        phaseEndTime = stateDTO.phaseEndTime,
        round = stateDTO.round,
        turn = stateDTO.turn,
        winner = stateDTO.winner
    )

    companion object {
        private const val PHASE_REGEX = "WAITING_FOR_PLAYERS|GRID_LAYOUT|IN_PROGRESS|FINISHED"
        private const val MIN_ROUND = 1
    }
}

File: code/js/src/Services/services/games/models/games/GameStateModel.ts
/**
 * The Game State Model.
 *
 * @property phase the phase of the game
 * @property phaseEndTime the time when the current phase ends
 * @property round the round of the game
 * @property turn the turn of the game
 * @property winner the winner of the game
 * @property endCause the cause of the game ending
 */
interface GameStateModel {
    phase: GamePhase
    phaseEndTime: number
    round: number | null
    turn: string | null
    winner: string | null
    endCause: string | null
}

type GamePhase = "WAITING_FOR_PLAYERS" | "DEPLOYING_FLEETS" | "IN_PROGRESS" | "FINISHED"
